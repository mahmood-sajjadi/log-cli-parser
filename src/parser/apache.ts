import type { Parser, AccessLog } from './types';

export const ApacheLogLineParser = (line: string): AccessLog => {
  const logPattern = /^(\S+) (\S+) (\S+) \[([^\]]+)\] "(\S+)\s(\S+)\s(\S+)" (\d{3}) (\d+|-) "[^"]+" "([^"]+)"/;

  const match = line.match(logPattern);
  if (!match) {
    throw new Error("Invalid log format");
  }

  const [
    // eslint-disable-next-line @typescript-eslint/no-unused-vars
    _,
    ip,
    logName,
    user,
    datetime,
    method,
    uri,
    protocol,
    status,
    size,
    agent,
  ] = match;

  return {
    accessTime: new Date(datetime),
    agent,
    ip,
    remoteLogName: logName,
    Request: {
      method,
      protocol,
      uri,
    },
    responseSize: size === '-' ? '-' : Number.parseInt(size, 10),
    statusCode: Number.parseInt(status, 10),
    user,
  }
}

/**
 * @deprecated Use `ApacheLogLineParser` instead. This function will be removed in version 2.0.
 */
export const ApacheLogFileParserFactory = (delimiter = '\n') => {
  const ApacheLogFileParser: Parser = (data: string) => {
    return data.split(delimiter).map(line => ApacheLogLineParser(line));
  }
  return ApacheLogFileParser;
}
