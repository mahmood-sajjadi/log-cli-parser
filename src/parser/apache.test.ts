import { ApacheLogLineParser, ApacheLogFileParserFactory } from './apache';

describe('ApacheLogLineParser', () => {
  test('WHEN line match PARSE line', () => {
    const line = '177.71.128.21 - - [10/Jul/2018:22:21:28 0200] "GET /intranet-analytics/ HTTP/1.1" 200 3574 "-" "Mozilla/5.0 (X11; U; Linux x86_64; fr-FR) AppleWebKit/534.7 (KHTML, like Gecko) Epiphany/2.30.6 Safari/534.7"';
    expect(ApacheLogLineParser(line)).toMatchSnapshot();
  });

  test('WHEN line not match THROW error', () => {
    expect(() => ApacheLogLineParser('')).toThrow('Invalid log format');
  });
});

describe('ApacheLogFileParser', () => {
  test('WHEN multi line match PARSE all lines', () => {
    const line = '177.71.128.21 - - [10/Jul/2018:22:21:28 0200] "GET /intranet-analytics/ HTTP/1.1" 200 3574 "-" "Mozilla/5.0 (X11; U; Linux x86_64; fr-FR) AppleWebKit/534.7 (KHTML, like Gecko) Epiphany/2.30.6 Safari/534.7"';
    const delimiter = '\n';
    const ApacheLogFileParser = ApacheLogFileParserFactory(delimiter);
    const result = ApacheLogFileParser(line + delimiter + line);
    expect(result).toHaveLength(2);
    expect(result).toMatchSnapshot();
  });
});
