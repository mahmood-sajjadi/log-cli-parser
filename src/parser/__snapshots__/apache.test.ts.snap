// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ApacheLogFileParser WHEN multi line match PARSE all lines 1`] = `
[
  {
    "Request": {
      "method": "GET",
      "protocol": "HTTP/1.1",
      "uri": "/intranet-analytics/",
    },
    "accessTime": Date { NaN },
    "agent": "Mozilla/5.0 (X11; U; Linux x86_64; fr-FR) AppleWebKit/534.7 (KHTML, like Gecko) Epiphany/2.30.6 Safari/534.7",
    "ip": "177.71.128.21",
    "remoteLogName": "-",
    "responseSize": 3574,
    "statusCode": 200,
    "user": "-",
  },
  {
    "Request": {
      "method": "GET",
      "protocol": "HTTP/1.1",
      "uri": "/intranet-analytics/",
    },
    "accessTime": Date { NaN },
    "agent": "Mozilla/5.0 (X11; U; Linux x86_64; fr-FR) AppleWebKit/534.7 (KHTML, like Gecko) Epiphany/2.30.6 Safari/534.7",
    "ip": "177.71.128.21",
    "remoteLogName": "-",
    "responseSize": 3574,
    "statusCode": 200,
    "user": "-",
  },
]
`;

exports[`ApacheLogLineParser WHEN line match PARSE line 1`] = `
{
  "Request": {
    "method": "GET",
    "protocol": "HTTP/1.1",
    "uri": "/intranet-analytics/",
  },
  "accessTime": Date { NaN },
  "agent": "Mozilla/5.0 (X11; U; Linux x86_64; fr-FR) AppleWebKit/534.7 (KHTML, like Gecko) Epiphany/2.30.6 Safari/534.7",
  "ip": "177.71.128.21",
  "remoteLogName": "-",
  "responseSize": 3574,
  "statusCode": 200,
  "user": "-",
}
`;
